### PR TITLE
RUMM-1691 Reduce utilization of `CTTelephonyNetworkInfo` to mitigate `CarrierInfoProvider` crash

### PR DESCRIPTION
**Pull request by ncreated**
### What and why?

📦 This PR updates the way we retrieve `CarrierInfo` information on iOS 12+. It's an educated guess to mitigate infrequent (2 impacted users) crashes reported in #623 and #619. Instead of pulling `CTTelephonyNetworkInfo` for each collected event to create `CarrierInfo`, now we only listen to `CTTelephonyNetworkInfo` changes to update `CarrierInfo` when necessary.

We use `CTTelephonyNetworkInfo` (from `CoreTelephony` framework) to read carrier information:

Screenshot 2021-10-07 at 12 46 30

A loop of investigation, research and debugging, lead me to following conclusions:
* C1. Reported crashes are very infrequent (one report on iOS 14.8 and one on iOS 14.6),
* C2. I observed is that `CTTelephonyNetworkInfo` only changes when swapping SIM card to another one (it does not change when removing SIM card - the last known info is still available). This is the only example listed in [Apple doc](https://developer.apple.com/documentation/coretelephony/cttelephonynetworkinfo).
* C3. We already had similar issues with utilization of Apple networking frameworks:
   * [recent crash on polling `ProcessInfo.isLowPowerModeEnabled`](https://github.com/DataDog/dd-sdk-ios/pull/613),
   * [old issue with pulling `NWPath.currentPath `](https://github.com/DataDog/dd-sdk-ios/pull/109).

Both issues in C3 were fixed by changing from pulling model, to subscription model (it was also a [recommendation I received](https://developer.apple.com/forums/thread/133112) for `NWPath` issue from Apple's engineer). Given that `CTTelephonyNetworkInfo` changes are very infrequent (C2) it makes no sense to pull this value and compute `Datadog.CarrierInfo` for each event we collect. Instead, we can listen to `CTTelephonyNetworkInfo` changes and update `Datadog.CarrierInfo` only when necessary.

I can't confirm if this fixes the crash, because I didn't manage to reproduce it even when stress testing our SDK with `~500logs/s` sent from iPhone X and swapping between different SIM cards during the test.

Screenshot 2021-10-06 at 18 27 04

### How?

Now, we use [`serviceSubscriberCellularProvidersDidUpdateNotifier`](https://developer.apple.com/documentation/coretelephony/cttelephonynetworkinfo/3024512-servicesubscribercellularprovide) available on iOS 12+ to update `CarrierInfo`. Initial value is read when SDK starts.

The logic for iOS 11 remains the same with the difference that now we also read initial value for `CarrierInfo` when starting the SDK. Although there is an API for listening to `CTTelephonyNetworkInfo` updates on iOS 11, it is deprecated (ref.: [`CTRadioAccessTechnologyDidChange `](https://developer.apple.com/documentation/foundation/nsnotification/name/1616908-ctradioaccesstechnologydidchange)) and because we have no reports for iOS 11, I decided to not change existing logic.

Last, I added simple utility to stress test Logging feature in Example app:

Screenshot 2021-10-07 at 12 42 58

### Review checklist

- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference

### CODE REVIEW

File: Sources/Datadog/Core/System/CarrierInfoProvider.swift [hunk 1]
```diff
@@ -66,88 +66,145 @@ extension CarrierInfo.RadioAccessTechnology {
...
+
+        // The `serviceSubscriberCellularProvidersDidUpdateNotifier` block object executes on the default priority
+        // global dispatch queue when the user’s cellular provider information changes.
+        // This occurs, for example, if a user swaps the device’s SIM card with one from another provider, while the app is running.
+        // ref.: https://developer.apple.com/documentation/coretelephony/cttelephonynetworkinfo/3024512-servicesubscribercellularprovide
+        networkInfo.serviceSubscriberCellularProvidersDidUpdateNotifier = { [weak self] cellularProviderKey in
+            let carrierInfo = iOS12CarrierInfoProvider.readCarrierInfo(
+                from: networkInfo,
+                cellularProviderKey: networkInfo.serviceCurrentRadioAccessTechnology?.keys.first
+            )
```
Comment (ncreated) [inline Sources/Datadog/Core/System/CarrierInfoProvider.swift:136]: I've been debugging what happens to `networkInfo: CTTelephonyNetworkInfo` when removing and changing SIM card. The object itself is not deallocated, only its values are updated:
Screenshot 2021-10-06 at 18 43 46

What changes though is the `CTCarrier` retrieved in `readCarrierInfo(from:cellularProviderKey:)`:
```swift
let carrier: CTCarrier = networkInfo.serviceSubscriberCellularProviders?[cellularProviderKey]
```
Screenshot 2021-10-07 at 13 40 44


The `swift_unknownObjectRetain` in reported crash:
```
0  libswiftCore.dylib             0x18ed94b70 swift_unknownObjectRetain + 20
1  libswiftFoundation.dylib       0x18e8b7f48 closure #2 in static Dictionary._forceBridgeFromObjectiveC(_:result:) + 736
2  libswiftFoundation.dylib       0x18e8b8000 partial apply for closure #2 in static Dictionary._forceBridgeFromObjectiveC(_:result:) + 36
3  libswiftFoundation.dylib       0x18e8bd32c _NativeDictionary.init(_unsafeUninitializedCapacity:allowingDuplicates:initializingWith:) + 340
4  libswiftFoundation.dylib       0x18e8b6ad0 static Dictionary._unconditionallyBridgeFromObjectiveC(_:) + 1368
5  Datadog                        0x103c948b8 iOSCarrierInfoProvider.current.getter + 132 (CarrierInfoProvider.swift:132)
```
seems to correspond to extending the `CTCarrier's` lifespan in our code after it was released in `CoreTelephony`. This is one of the scenarios I've been trying to reproduce (extending it explicitly), but it didn't lead to any crash.

Comment (maxep) [reply Sources/Datadog/Core/System/CarrierInfoProvider.swift:136]: This creates a retain cycle of `networkInfo`.

Comment (maxep) [reply Sources/Datadog/Core/System/CarrierInfoProvider.swift:136]: Why not using the `cellularProviderKey` from the `serviceSubscriberCellularProvidersDidUpdateNotifier` callback?

File: Sources/Datadog/Core/System/CarrierInfoProvider.swift [hunk 2]
```diff
@@ -66,88 +66,145 @@ extension CarrierInfo.RadioAccessTechnology {
...
+
+    init(networkInfo: CTTelephonyNetworkInfo) {
+        self.networkInfo = networkInfo
+        self.publisher = ValuePublisher(
+            initialValue: iOS11CarrierInfoProvider.readCarrierInfo(from: networkInfo)
         )
     }
+
+    private static func readCarrierInfo(from networkInfo: CTTelephonyNetworkInfo) -> CarrierInfo? {
+        if let radioTechnology = networkInfo.currentRadioAccessTechnology,
```
Comment (maxep) [inline Sources/Datadog/Core/System/CarrierInfoProvider.swift:182]: /nit using a `guard` would make it more readable IMHO 👓

File: Sources/Datadog/Core/System/CarrierInfoProvider.swift (hunk 1 above)
Comment (ncreated) [reply Sources/Datadog/Core/System/CarrierInfoProvider.swift:136]: > I see a retain cycle of networkInfo here.

Oh yeah, my bad. It doesn't make any bad side effects, as `networkInfo` is a singleton (once obtained from `CoreTelephony` it never changes). Anyway, will fix that, good catch 👍 .

> Also, why not using the `cellularProviderKey` returned by the `serviceSubscriberCellularProvidersDidUpdateNotifier` closure?

My bad, again, good catch 👍. In all my tests, it has the same value as `serviceCurrentRadioAccessTechnology's` `.keys.first`, but the entire point for having this callback is to indeed use `cellularProviderKey` instead. Will fix.

Comment (ncreated) [reply Sources/Datadog/Core/System/CarrierInfoProvider.swift]: Fixed 👍. I did one more test, it works.

Screenshot 2021-10-08 at 16 05 05

File: Sources/Datadog/Core/System/CarrierInfoProvider.swift [hunk 3]
```diff
@@ -66,88 +66,147 @@ extension CarrierInfo.RadioAccessTechnology {
...
+           let carrier = networkInfo.serviceSubscriberCellularProviders?[cellularProviderKey] else {
+            return nil // the service is not registered on any network
         }
+        return CarrierInfo(
+            carrierName: carrier.carrierName,
+            carrierISOCountryCode: carrier.isoCountryCode,
+            carrierAllowsVOIP: carrier.allowsVOIP,
+            radioAccessTechnology: .init(ctRadioAccessTechnologyConstant: radioTechnology)
+        )
+    }
```
Comment (maxep) [inline Sources/Datadog/Core/System/CarrierInfoProvider.swift:159]: /nit I think you can simplify that by using an instance method to read the carrier, something like:

```swift
init(networkInfo: CTTelephonyNetworkInfo) {
    ...
    networkInfo.serviceSubscriberCellularProvidersDidUpdateNotifier = { [weak self] cellularProviderKey in
        let carrierInfo = self?.readCarrierInfo(provider: cellularProviderKey)
        self?.publisher.publishAsync(carrierInfo)
    }
}

private func readCarrierInfo(provider key: String) -> CarrierInfo? {
    guard 
       let radioTechnology = networkInfo.serviceCurrentRadioAccessTechnology?[key],
       let carrier = networkInfo.serviceSubscriberCellularProviders?[key] 
    else {
       return nil // the service is not registered on any network
    }
    return CarrierInfo(
          carrierName: carrier.carrierName,
          carrierISOCountryCode: carrier.isoCountryCode,
          carrierAllowsVOIP: carrier.allowsVOIP,
          radioAccessTechnology: .init(ctRadioAccessTechnologyConstant: radioTechnology)
    )
}

**Review by maxep — APPROVED** [review]
LGTM 👌 
I've just a minor suggestion

**Review by buranmert — APPROVED** [review]

**Pull request merged**